engines/leveldb/leveldb_test.go

```diff
@@ -0,0 +1,235 @@
+func checkDelete(ldbStore engine.Store, opts *options.Options, key []byte, t *testing.T) {
+	err := ldbStore.Delete(key, opts)
+	require.NoError(t, err)
+
+	value, err := ldbStore.Get(key, opts)
+	require.Equal(t, err, engine.ErrNotFound)
+	require.Empty(t, value)
+}
+
+func TestLeveldbEngine(t *testing.T) {
+	// Setup a levelDB Engine.
+	ldbEngine, ldbPath := setupLeveldbEngine(t)
+	require.Equal(t, "leveldb", ldbEngine.Engine())
+
+	// Ensure the db was created.
+	require.DirExists(t, ldbPath)
+
+	// Teardown after finishing the test.
+	defer os.RemoveAll(ldbPath)
+	defer ldbEngine.Close()
+
+	// Use a constant key to ensure namespaces
+	// are working correctly.
+	key := []byte("foo")
+
+	// Check Put, Get and Delete with a nil namespace.
+	value := []byte("nil")
+	checkPut(ldbEngine, nil, key, value, t)
+	checkGet(ldbEngine, nil, key, value, t)
+	checkDelete(ldbEngine, nil, key, t)
+
+	// Iterate through a list of namespaces and ensure
+	// Put, Get and Delete are working.
+	for _, namespace := range getNamespaces() {
+		opts := namespaceOpts(namespace, t)
+		value := []byte(namespace)
+		checkPut(ldbEngine, opts, key, value, t)
+		checkGet(ldbEngine, opts, key, value, t)
+		checkDelete(ldbEngine, opts, key, t)
+	}
+}
+
+func TestLeveldbTransactions(t *testing.T) {
+	ldbEngine, ldbPath := setupLeveldbEngine(t)
+
+	// Teardown after finishing the test
+	defer os.RemoveAll(ldbPath)
+	defer ldbEngine.Close()
+
+	// Use a constant key to ensure namespaces
+	// are working correctly.
+	key := []byte("tx")
+
+	// Iterate through a list of namespaces and ensure
+	// Put, Get and Delete are working with transactions.
+	for _, namespace := range getNamespaces() {
+		// Start a transaction with readonly set to false.
+		tx, err := ldbEngine.Begin(false)
+		require.NoError(t, err)
+
+		opts, err := options.New(options.WithNamespace(namespace))
+		require.NoError(t, err)
+		value := []byte(namespace)
+		checkPut(tx, opts, key, value, t)
+		checkGet(tx, opts, key, value, t)
+		checkDelete(tx, opts, key, t)
+
+		// Complete the transaction.
+		require.NoError(t, tx.Finish())
+	}
+
+	// Begin a transaction with readonly set to true.
+	tx, err := ldbEngine.Begin(true)
+	require.NoError(t, err)
+
+	// Ensure Put fails with readonly set.
+	err = tx.Put([]byte("bad"), []byte("write"), nil)
+	require.Equal(t, err, engine.ErrReadOnlyTx)
+	err = tx.Delete([]byte("bad"), nil)
+	require.Equal(t, err, engine.ErrReadOnlyTx)
+
+	// Complete the transaction.
+	require.NoError(t, tx.Finish())
+}
+
+func TestLevelDBIter(t *testing.T) {
+	ldbEngine, ldbPath := setupLeveldbEngine(t)
+
+	// Teardown after finishing the test
+	defer os.RemoveAll(ldbPath)
+	defer ldbEngine.Close()
+
+	for _, namespace := range getNamespaces() {
+		// TODO: figure out what to do with this testcase.
+		// Iter currently grabs the namespace by splitting
+		// on :: and grabbing the first string, so it only
+		// grabs "namespace".
+		if namespace == "namespace::with::colons" {
+			continue
+		}
```

Comment:
When you create the story for this (see other comment), please also reference this test as needing to be fixed!

Reply:
Done!